### PR TITLE
Fix gradcheck failing for OpInfo with TensorList inputs

### LINKED ISSUE #51996: OpInfo tests fail gradcheck for ops with TensorList inputs.
For ops like ``torch.stack`` and ``torch.linalg.multi_dot`` that take a ``Tensor[]`` as input as opposed to a single Tensor, OpInfo testing does not work properly.

If the `sample_inputs_func` returns a tuple of tensors per SampleInput, it will be passed as individual positional parameters to the op. If instead each SampleInput is returned with a list, then gradcheck will fail with the following error:

```
FAILED test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_multi_dot_cpu_float64 - ValueError: gradcheck expects at least one input tensor to require gradient, but none of the them have requires_grad=True.
```

The current workaround is to create a lambda as follows:
```
op=lambda *args, out=None: torch.linalg.multi_dot([*args], out=out),
```

However, this is changing the actual op which can cause problems in the future, for instance, when comparing the results of the op against NumPy.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @jianyuh @mruberry @heitorschueroff @walterddr @IvanYashchuk @VitalyFedyunin

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53418 Fix gradcheck failing for OpInfo with TensorList inputs**

This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

### CODE REVIEW

File: test/test_ops.py
Comment (albanD) [inline test/test_ops.py:71]: nit: is it important that inputs is a list instead of a tuple here?

**Review by albanD — APPROVED** [review]
lgtm

File: test/test_ops.py
Comment (heitorschueroff) [reply test/test_ops.py:71]: Nope, I made it a tuple. Thanks!

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:89]: This is probably fine. Just fyi I recently wrote a similar function:

https://github.com/pytorch/pytorch/blob/2d36b30a8c35cc5640e1c735583c105b7037f9ca/test/test_ops.py#L466

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:186]: What about making this None and if None it just points to the op?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:947]: I'm not sure if this PR can remove this class. cc @IvanYashchuk, you're updating this class in another PR of yours, right?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2521]: Let's file and link an issue for this

**Comment by mruberry** [comment]
I really like this change and the design looks good. I've asked @IvanYashchuk to weigh in on removing the two OpInfo classes for linalg operations. Although this PR can replace them in the current build, I think he may have plans for them.

Looks like the test failures are just mypy issues.

File: torch/testing/_internal/common_methods_invocations.py
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py:947]: It's great that `grad_func=` replaces `HermitianOpInfo` and `TriangularOpInfo`! But it will hit a merge conflict with https://github.com/pytorch/pytorch/pull/51107
Let's wait until that PR is merged and then rebase this one.

**Comment by IvanYashchuk** [comment]
This PR resolves https://github.com/pytorch/pytorch/issues/50837.

File: torch/testing/_internal/common_methods_invocations.py
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:89]: I moved this method to testing._internal.common_utils to be reused in common_methods_invocation and test_ops

File: torch/testing/_internal/common_methods_invocations.py
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:186]: It has to take op has its first argument so it can wrap the op variants as well. I gave it a default lambda as a way to document its use.

File: torch/testing/_internal/common_methods_invocations.py
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:2521]: Created issue here https://github.com/pytorch/pytorch/issues/53906

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:186]: OK.

File: torch/testing/_internal/common_methods_invocations.py
Comment (heitorschueroff) [inline torch/testing/_internal/common_methods_invocations.py:609]: For now, TensorList inputs must be List[Tensor]. @heitorschueroff find out why Tuple[Tensor] is not supported here.

**Pull request closed without merging**
